### PR TITLE
Adding KimiK2.5 ViT

### PR DESCRIPTION
**Pull request by depksingh**
## Motivation

As part of onboarding Kimi-K2.5 on sglang-jax https://github.com/sgl-project/sglang-jax/issues/1011, adding this PR on top the previous LLM backbone PR  #1331, to enable the vision capability of the model.

## Modifications

- Added ViT class for Kimi-K2.5
- Added vision model configs for the same

## Accuracy Tests

Launch command with v7x 2x2x2 topology with tp=32
```
JAX_COMPILATION_CACHE_DIR=/tmp/tpu_logs python3 -u -m sgl_jax.launch_server     --multimodal     --model-path=/local/kimi     --log-requests --trust-remote-code --page-size 256 --chunked-prefill-size 512 --max-running-requests 64 --tp-size 32 --ep-size 32 --precompile-bs-paddings 64
```

Tested and verified the model implementation by sending images to the model to identify the images:

*Input image & prompts*:

image

image

<br>
image
￼

image

image

image

## Accuracy Validation - MMMU 

Pasted Graphic 6
Pasted Graphic 7

OVERALL

## Benchmarking and Profiling
NA, will be part of E2E run

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [x] (Optional) The necessary documentation update.

### CODE REVIEW

**Comment by depksingh** [comment]
Hey @cjx0709 , @Bob-Chen222 , @chaitanyapk, Added the above PR for ViT for KimiK2.5, can you please review it whenever you get time.

**Comment by cjx0709** [comment]
@depksingh  hello, thanks for your contribution, can you also post the serve launch command and the hardware environment information.

**Comment by depksingh** [comment]
> @depksingh hello, thanks for your contribution, can you also post the serve launch command and the hardware environment information.

Ah, my bad, updated the description with the same.

File: python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py [hunk 1]
```diff
@@ -0,0 +1,765 @@
...
+        q = apply_2d_rope(q, cos_emb, sin_emb)
+        k = apply_2d_rope(k, cos_emb, sin_emb)
+
+        align_seq_len = align_to(sum_seq_len, 256)
+
+        pad_q = q
+        pad_k = k
+        pad_v = v
+
+        segment_ids = None
```
Comment (Bob-Chen222) [inline python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py]: It looks like segment_ids is only initalized when there is padding. But what if two images need to be isolated?

File: python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+
+
+class KimiK25VisionBlock(nnx.Module):
+
+    def __init__(
+        self,
+        config: KimiK25ModelVitConfig,
+        dtype: jnp.dtype,
+        mesh: Mesh | None = None,
+        norm_eps: float = 1e-6,
```
Comment (Bob-Chen222) [inline python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py:305]: seems this norm_eps is never used. Did you run the pre-commit?

File: python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+        dtype: jnp.dtype = jnp.bfloat16,
+        rngs: nnx.Rngs | None = None,
+        mesh: Mesh | None = None,
+    ) -> None:
+
+        self.config = config
+        self.dtype = dtype
+        self.mesh = mesh
+
+        self.vision_tower = VisionTower(config, dtype, rngs, mesh, config.projector_ln_eps)
```
Comment (Bob-Chen222) [inline python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py]: looks like projector_ln_eps is also never used

File: python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+    out = np.einsum("m,d->md", pos, omega)
+
+    emb_sin = np.sin(out)
+    emb_cos = np.cos(out)
+
+    emb = np.concatenate([emb_sin, emb_cos], axis=1)
+    return emb
+
+
+def get_1d_sincos_pos_embed(embed_dim, t_size, cls_token=False):
```
Comment (Bob-Chen222) [inline python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py]: is this function being used anywhere? if not probably can be removed

File: python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+        dtype: jnp.dtype,
+        mesh: Mesh | None = None,
+        rngs: nnx.Rngs | None = None,
+    ):
+        in_features = config.vt_hidden_size
+        intermediate_size = config.vt_intermediate_size
+
+        _rngs = rngs or nnx.Rngs(0)
+
+        self.mesh = mesh
```
Comment (Bob-Chen222) [inline python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py]: looks like mesh is not used in MLP

**Comment by lianfang0320** [comment]
@depksingh This PR seems to depend on the Kimi pipeline wiring in [#1314](https://github.com/sgl-project/sglang-jax/pull/1314), which is not merged into main yet. Could you either land/update #1314 first, then rebase this PR on top of it, or include the required pipeline wiring in this PR? Right now #1388 against main only adds the ViT/config files, so the submitted launch command is not reproducible from this PR alone.

What do you think? @cjx0709 @depksingh

**Comment by depksingh** [comment]
> @depksingh This PR seems to depend on the Kimi pipeline wiring in [#1314](https://github.com/sgl-project/sglang-jax/pull/1314), which is not merged into main yet. Could you either land/update #1314 first, then rebase this PR on top of it, or include the required pipeline wiring in this PR? Right now #1388 against main only adds the ViT/config files, so the submitted launch command is not reproducible from this PR alone.
> 
> What do you think? @cjx0709 @depksingh

Hey @lianfang0320 , yeah, so there is one more PR that is open i.e. regarding the INT4 quantization of Kimi-k2.5 model https://github.com/sgl-project/sglang-jax/pull/1334, which adds the support for INT4 quant for the `moonshot-ai/kimi-k2.5` model. I'm working with converted bf16 weights model for now, with which I tested both the text_generation here https://github.com/sgl-project/sglang-jax/pull/1331 and this ViT PR as well. so even with merging the KIMI pipeline PR (i.e. #1314 ), the server launch command will need the converted bf16 weights to run and not work with the native INT4 weights. Which would be resolved when the #1334 PR is merged. Our idea was to merge PRs in this Order
1. VL_Generation (Tested) -> 2. ViT (Tested) -> 3. Quantization PR (Tested E2E) 4. E2E pipeline 

After the 4th PR all the changes can be run e2e with the server launch command is what we were thinking. That's why we followed the same pattern as LLM backbone PR (#1331) i.e. completed the implementation of the ViT components and tested it e2e in isolation, instead of adding e2e pipeline changes without INT4 to avoid  showing `moonshotai/Kimi-K2.5` INT4 is supported even though the INT4 changes aren't fully integrated. That is the reason if you see in the server launch command we are using the model path as /local/kimi, as these are de-quantized bf16 weights. 
Please let me know what do you think if this isolated PR for vit will work or should we first merge the Quantization PR, followed by the pipeline and then this PR or if you have any other suggestions.

File: python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py (hunk 1 above)
Comment (depksingh) [reply python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py]: Nice catch, fixed it to handle incase of no padding as well.

File: python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py (hunk 4 above)
Comment (depksingh) [reply python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py]: good catch, this was specifcally for handling frames in videos to have the PE for the temporal axis, will add it once when adding video support

File: python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py (hunk 5 above)
Comment (depksingh) [reply python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py]: yes, my bad, removed it from the VisionMLP class, as vision is using it so being passed in vision block.

File: python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py (hunk 2 above)
Comment (depksingh) [reply python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py:305]: yes, seems like pre-commit didn't catch it as it passed when raising pr. removed the norm_eps as its using default value and passing the correct one for projector layer.

File: python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py (hunk 3 above)
Comment (depksingh) [reply python/sgl_jax/srt/multimodal/models/kimi_k25/kimi_k25_vit.py]: updated to use it in projection layer.

**Comment by depksingh** [comment]
> > @depksingh This PR seems to depend on the Kimi pipeline wiring in [#1314](https://github.com/sgl-project/sglang-jax/pull/1314), which is not merged into main yet. Could you either land/update #1314 first, then rebase this PR on top of it, or include the required pipeline wiring in this PR? Right now #1388 against main only adds the ViT/config files, so the submitted launch command is not reproducible from this PR alone.
> > What do you think? @cjx0709 @depksingh
> 
> Hey @lianfang0320 , yeah, so there is one more PR that is open i.e. regarding the INT4 quantization of Kimi-k2.5 model #1334, which adds the support for INT4 quant for the `moonshot-ai/kimi-k2.5` model. I'm working with converted bf16 weights model for now, with which I tested both the text_generation here #1331 and this ViT PR as well. so even with merging the KIMI pipeline PR (i.e. #1314 ), the server launch command will need the converted bf16 weights to run and not work with the native INT4 weights. Which would be resolved when the #1334 PR is merged. Our idea was to merge PRs in this Order
> 
> 1. VL_Generation (Tested) -> 2. ViT (Tested) -> 3. Quantization PR (Tested E2E) 4. E2E pipeline
> 
> After the 4th PR all the changes can be run e2e with the server launch command is what we were thinking. That's why we followed the same pattern as LLM backbone PR (#1331) i.e. completed the implementation of the ViT components and tested it e2e in isolation, instead of adding e2e pipeline changes without INT4 to avoid showing `moonshotai/Kimi-K2.5` INT4 is supported even though the INT4 changes aren't fully integrated. That is the reason if you see in the server launch command we are using the model path as /local/kimi, as these are de-quantized bf16 weights. Please let me know what do you think if this isolated PR for vit will work or should we first merge the Quantization PR, followed by the pipeline and then this PR or if you have any other suggestions.

Hey @lianfang0320 , would it be fine if I share launch server command with this model I found https://huggingface.co/ananayarora/Kimi-K2.5-BF16 (It seems to be the correct weights which I checked after prompts and accuracy checks) which is a de-quantized bf16 variant which is manually decompressed. It would unblock us till the INT4 PR is merged, for verifying the server launch and e2e verification on your end, or we can wait for it to merge. Please let me know your views.

**Review by Bob-Chen222 — APPROVED** [review]
LGTM. cc @cjx0709 @JamesBrianD  if you have any concerns

**Comment by Bob-Chen222** [comment]
close and reopen for triggering CI
